### PR TITLE
Using BuiltinClusterDataSerializable for Buffer, JsonObject and JsonArray instead of Refection implementation

### PR DESCRIPTION
**Pull request by dinhnn**
Almost message send via vertx cluster using JsonObject, JsonArray & Buffer.
Implement using BuiltinClusterDataSerializable will improve performance of ~~eventbus~~ AsyncMap
Here is a summary of changes in this PR:
- BufferDataSerializable
- JsonDataSerializable
- JsonArraySerializable

### CODE REVIEW

**Comment by cescoffier** [comment]
Did you measure the performance improvements ?

File: src/main/java/io/vertx/spi/cluster/hazelcast/impl/HazelcastAsyncMap.java [hunk 1]
```diff
@@ -16,189 +16,208 @@
...
-    public int hashCode() {
-      return clusterSerializable != null ? clusterSerializable.hashCode() : 0;
-    }
-
-    public ClusterSerializable clusterSerializable() {
-      return clusterSerializable;
-    }
-  }
-
+	private final Vertx vertx;
```
Comment (purplefox) [inline src/main/java/io/vertx/spi/cluster/hazelcast/impl/HazelcastAsyncMap.java]: indentation should be 2 chars

**Comment by purplefox** [comment]
This PR has changed the indentation from the normal 2 chars to 4 on every file it has touched making it very hard to see the actual changes. @dinhnn  please change indentation to 2 chars

File: src/main/java/io/vertx/spi/cluster/hazelcast/impl/serializable/BuiltinClusterDataSerializable.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+		
+	}
+	@Override
+	public int getFactoryId() {
+		return VertxDataSerializerHook.FACTORY_ID;
+	}
+	@Override
+	public void writeData(ObjectDataOutput objectDataOutput) throws IOException {
+		objectDataOutput.writeUTF(clusterSerializable.getClass().getName());
+    Buffer buffer = Buffer.buffer();
```
Comment (purplefox) [inline src/main/java/io/vertx/spi/cluster/hazelcast/impl/serializable/BuiltinClusterDataSerializable.java]: indentation messed up - mixing spaces and tabs?

File: src/main/java/io/vertx/spi/cluster/hazelcast/impl/serializable/VertxDataSerializableFactory.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+public class VertxDataSerializableFactory implements DataSerializableFactory {
+	@Override
+	public IdentifiedDataSerializable create(int id) {
+		switch (id) {
+		case 0:
+			return new BufferDataSerializable();
+		case 1:
+			return new JsonDataSerializable();
+		default:
```
Comment (purplefox) [inline src/main/java/io/vertx/spi/cluster/hazelcast/impl/serializable/VertxDataSerializableFactory.java]: be explicit and use 2 for JsonArray. default should throw an exception

File: src/main/java/io/vertx/spi/cluster/hazelcast/impl/serializable/VertxDataSerializerHook.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package io.vertx.spi.cluster.hazelcast.impl.serializable;
+
+import com.hazelcast.nio.serialization.DataSerializableFactory;
+
+public class VertxDataSerializerHook implements com.hazelcast.internal.serialization.DataSerializerHook {
+	public final static int FACTORY_ID = 2906;
```
Comment (purplefox) [inline src/main/java/io/vertx/spi/cluster/hazelcast/impl/serializable/VertxDataSerializerHook.java]: why 2906?

**Comment by purplefox** [comment]
I think this would also introduce an incompatible change across the cluster.

**Comment by cescoffier** [comment]
I agree with @purplefox 
I'm a bit concerned with the compatibility with application using the previous format (for a cluster using the same HZ version, for example 3.6.1, clients without this and client with this should still be able to interact).

**Comment by purplefox** [comment]
+1 If there can be shown to be a performance improvement I think there is value in the PR, but not as the default behaviour (it could be enable using a system property for example)

File: src/main/java/io/vertx/spi/cluster/hazelcast/impl/HazelcastAsyncMap.java (hunk 1 above)
Comment (dinhnn) [inline src/main/java/io/vertx/spi/cluster/hazelcast/impl/HazelcastAsyncMap.java]: Yes, I will reformat it

**Comment by dinhnn** [comment]
Now I'm busy, but I will create performance test for it, please wait

**Comment by tsegismont** [comment]
Reopen if needed.

**Pull request closed without merging**